zkevm-circuits/src/pi_circuit2.rs

```diff
@@ -0,0 +1,3865 @@
+                blk_hdr_rlp_len_calc,
+                Rotation(
+                    (BASE_FEE_RLP_OFFSET + BASE_FEE_SIZE - 3)
+                        .try_into()
+                        .unwrap(),
+                ),
+            );
+
+            // For the block header, the total RLP length is always two bytes long and only
+            // the LSB fluctuates: Minimum total length: lengths of all the
+            // fixed size fields + all the RLP headers = 527 bytes (0x020F)
+            // Maximum total length: minimum total length + (maximum length of variable zize
+            // field) = 527 + 4*32+1*8 = 663 (0x0297) Actual total length:
+            // minimum total length + length of all variable size fields (number, gas_limit,
+            // gas_used, timestamp, base fee)
+            cb.require_equal(
+                "LSB(total_len) = min(LSB(total_len)) + sum(Len(<var field>))",
+                total_len,
+                0x0F.expr()
+                    + number_len
+                    + gas_limit_len
+                    + gas_used_len
+                    + timestamp_len
+                    + base_fee_len,
+            );
+
+            cb.gate(meta.query_selector(q_blk_hdr_total_len))
+        });
+
+        // Reconstruct field values
+        for selector in [
+            q_parent_hash,
+            q_beneficiary,
+            q_state_root,
+            q_transactions_root,
+            q_receipts_root,
+            q_number,
+            q_gas_limit,
+            q_gas_used,
+            q_timestamp,
+            q_mix_hash,
+            q_base_fee_per_gas,
+            q_withdrawals_root,
+        ] {
+            meta.create_gate(
+                "Block header RLP: reconstructing header field values from RLP",
+                |meta| {
+                    let mut cb = BaseConstraintBuilder::new(MAX_DEGREE);
+
+                    let q_hi = meta.query_selector(q_hi);
+                    let q_lo_cur = meta.query_fixed(q_lo, Rotation::cur());
+                    let q_lo_prev = meta.query_fixed(q_lo, Rotation::prev());
+
+                    let selector = meta.query_fixed(selector, Rotation::cur());
+                    let blk_hdr_rlp = meta.query_advice(blk_hdr_rlp, Rotation::cur());
+                    let blk_hdr_reconstruct_value_cur =
+                        meta.query_advice(blk_hdr_reconstruct_value, Rotation::cur());
+                    let blk_hdr_reconstruct_value_prev =
+                        meta.query_advice(blk_hdr_reconstruct_value, Rotation::prev());
+
+                    cb.condition(and::expr([selector.clone(), q_hi]), |cb| {
+                        cb.require_equal(
+                            "byte_hi[n]*2^8 + byte_hi[n+1]",
+                            blk_hdr_reconstruct_value_cur.clone(),
+                            blk_hdr_reconstruct_value_prev.clone() * 256.expr()
+                                + blk_hdr_rlp.clone(),
+                        )
+                    });
```

Comment:
Reconstructing the literal values is only used for checking that the values given in RLP are the correct ones. It can be replaced with RLC and save us ~10 lookups. Not sure how much it'll help performance, need your input @Brechtpd

Reply:
implemented the optimisation